### PR TITLE
Add UI template

### PR DESCRIPTION
**Pull request by pdesaulniers**
This PR adds support for generating plugin UI boilerplate.

It adds a `ui_type` variable to the cookiecutter.json file, which can be either `opengl`, `nanovg` or `none`.

### CODE REVIEW

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/DistrhoPluginInfo.h [hunk 1]
```diff
@@ -21,12 +21,20 @@
...
+{% if cookiecutter.ui_type == "none" %}
+#define DISTRHO_PLUGIN_HAS_UI        0
+{% else %}
+#define DISTRHO_PLUGIN_HAS_UI        1
+{%- if cookiecutter.ui_type == "nanovg" %}
+#define DISTRHO_UI_USE_NANOVG        1
+{%- else %}
+#define DISTRHO_UI_USE_NANOVG        0
+{%- endif %}
+{% endif %}
```
Comment (SpotlightKid) [inline {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/DistrhoPluginInfo.h]: This could be shortened to:

    #define DISTRHO_PLUGIN_HAS_UI        {{ 0 if cookiecutter.ui_type == "none" else 1 }}
    #define DISTRHO_UI_USE_NANOVG        {{ 1 if cookiecutter.ui_type == "nanovg" else 0 }}

This leaves the `#define`s in, even if no UI is selected, but I think that's a good thing.

File: cookiecutter.json [hunk 2]
```diff
@@ -11,5 +11,6 @@
     "plugin_uri": "https://{{ cookiecutter.domain }}/plugins/{{ cookiecutter.repo_name }}",
     "project_license": "MIT",
     "version": "0.1.0",
-    "year": "2018"
+    "year": "2019",
```
Comment (SpotlightKid) [inline cookiecutter.json:14]: This should really be in another PR, but I want to automate updating this variable using the technique described in the link below, so leave it in or remove the commit, doesn't matter.

https://cookiecutter.readthedocs.io/en/latest/advanced/template_extensions.html

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile [hunk 3]
```diff
@@ -9,27 +9,61 @@
...
 NAME = {{ cookiecutter.plugin_name | lower }}
 
+# --------------------------------------------------------------
+# Plugin types to build
+
+BUILD_LV2 ?= true
+BUILD_VST2 ?= true
+BUILD_JACK ?= true
+BUILD_DSSI ?= false
+BUILD_LADSPA ?= false
```
Comment (SpotlightKid) [inline {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile]: I would like the LADSPA format to be enabled by default for the time being. There are still hosts, which do not support LV2. maybe like this?

     BUILD_LADSPA = {{ "false" if cookiecutter.ui_type == "none" else "true" }}

File: cookiecutter.json [hunk 4]
```diff
@@ -11,5 +11,6 @@
...
+    "ui_type": ["opengl", "nanovg", "none"]
```
Comment (SpotlightKid) [inline cookiecutter.json:15]: What about cairo? Is this usable yet?

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.hpp{%- endif %} [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+{% endif %}
+    bool onKeyboard(const KeyboardEvent& ev) override;
+    bool onSpecial(const SpecialEvent& ev) override;
+    bool onMouse(const MouseEvent& ev) override;
+    bool onMotion(const MotionEvent& ev) override;
+    bool onScroll(const ScrollEvent& ev) override;
+
+private:
+    float    fParams[Plugin{{ cookiecutter.plugin_name }}::paramCount];
+    double   fSampleRate;
```
Comment (SpotlightKid) [inline {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.hpp{%- endif %}]: Does the UI really need to keep track of its own copy of parameter values and the sampling rate? For a UI, that simply provides widgets to change parameters, I don't think so. If this is kept in, maybe some code using these variables should also be provided (other than `programLoaded` / `sampleRateChanged`)?

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %} [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        case Plugin{{ cookiecutter.plugin_name }}::paramVolume:
+            // do something when volume param is set
+            break;
+    }
+}
+
+void UI{{ cookiecutter.plugin_name }}::programLoaded(uint32_t index) {
+    switch (index) {
+        case 0:
+            fParams[PluginSimpleGain::paramVolume] = 0.1f;
```
Comment (SpotlightKid) [inline {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %}]: This seems a bit pointless on it's own. Maybe instead insert a comment like this:

    // Set values for each parameter and update their widgets

I'm not sure, would it be ok, to call `parameterChanged` here?

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %} [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "UI{{ cookiecutter.plugin_name }}.hpp"
+
+START_NAMESPACE_DISTRHO
+
```
Comment (SpotlightKid) [inline {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %}:21]: I'd like a few more method documentation comments in this file. Along the lines of https://github.com/DISTRHO/DPF/blob/master/examples/Meters/ExampleUIMeters.cpp#L50 ff.

**Review by SpotlightKid — CHANGES_REQUESTED** [review]
Thanks for this contribution! I like it very much, in general, but I've commented on a few things I would like to discuss or be changed before merging.

File: cookiecutter.json (hunk 4 above)
Comment (pdesaulniers) [reply cookiecutter.json:15]: According to falkTX, Cairo support in DPF is not finalized yet.

File: cookiecutter.json (hunk 2 above)
Comment (pdesaulniers) [reply cookiecutter.json:14]: I guess I'll leave it here for now.

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.hpp{%- endif %} (hunk 5 above)
Comment (pdesaulniers) [reply {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.hpp{%- endif %}]: This doesn't really need to be kept in. Besides, I don't think that keeping track of the sample rate is useful at all, as we can just call `getSampleRate()` at any moment.

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile (hunk 3 above)
Comment (pdesaulniers) [reply {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile]: Not sure. Did you mean the opposite? 
```
BUILD_LADSPA ?= {{ "true" if cookiecutter.ui_type == "none" else "false" }}
```
Instead, perhaps we could add some variables in the cookiecutter.json file to determine which formats to enable? Such as `build_ladspa`, `build_dssi`, etc.

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %} (hunk 6 above)
Comment (pdesaulniers) [reply {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %}]: Yes, I think calling parameterChanged makes sense here.

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %} (hunk 7 above)
Comment (pdesaulniers) [reply {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/{%- if cookiecutter.ui_type != "none" %}UI{{ cookiecutter.plugin_name }}.cpp{%- endif %}:21]: OK, I've added some documentation.

File: {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile (hunk 3 above)
Comment (SpotlightKid) [reply {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile]: Yes, sorry, you're right, of course. Only build LADSPA format if no GUI is used.

Comment (pdesaulniers) [reply {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile]: OK, done.

Comment (SpotlightKid) [reply {{ cookiecutter.repo_name }}/plugins/{{ cookiecutter.plugin_name }}/Makefile]: Unfortunately, cookiecutter does not have support for "pick-list" variables, only single-choice ones, so we need to ask whether to build each format separately. This sth. for another PR, though (or I'll do it myself9)

File: cookiecutter.json (hunk 4 above)
Comment (SpotlightKid) [reply cookiecutter.json:15]: Ok, support for it can be added later.

Comment (jpcima) [reply cookiecutter.json:15]: Cairo support is working, I have released a few plugins based on it.
What's not available on cairo: image, standard image-based widgets, geometric shape drawing.
None of these are vital when using custom widgets and features already in Cairo.
Personnally, I would appreciate if cairo support is added.
